tests/test_testserver.py

```diff
@@ -48,6 +48,47 @@ def test_context(self):
 
         time.sleep(0.1)  # Give server a moment to spin down
 
+    def test_server_disconnect_then_del_device_notification(self):
+        """Test no error thown, when ADS symbol with device_notification is cleaned up after the server went offline.
+
+        Tests fix of issue [#303](https://github.com/stlehmann/pyads/issues/303), original pull request: [#304](https://github.com/stlehmann/pyads/pull/304)
+        """
+        # 1. spin up the server
+        handler = BasicHandler()
+        test_server = AdsTestServer(handler=handler, logging=False)
+        test_server.start()
+        time.sleep(0.1)  # Give server a moment to spin up
+
+        # 2. open a plc connection to the test server:
+        plc = pyads.Connection(TEST_SERVER_AMS_NET_ID, TEST_SERVER_AMS_PORT)
+        plc.open()
+
+        # 3. add a variable, register a device notification with auto_update=True
+        test_int = pyads.AdsSymbol(plc, "TestSymbol", symbol_type=pyads.PLCTYPE_INT)
+        test_int.plc_type = pyads.PLCTYPE_INT
+        test_int.auto_update = True
+        time.sleep(0.1)  # Give server a moment
+
+        raised_error: str = ""
+
+        try:
+            # 4. stop the test server
+            test_server.stop()
+            time.sleep(0.1)  # Give server a moment
+
+            # some code, where test_int is cleared by the Garbage collector after the server was stopped
+            # (e.g. the machine with ADS Server disconnected)
+            # this raised an ADSError up to commit [a7af674](https://github.com/stlehmann/pyads/tree/a7af674b49b1c91966f2bac1f00f86273cbd9af8)
+            #  `clear_device_notifications()` failed, if not wrapped in try-catch as the server is no longer present.
+            test_int.__del__()
+        except pyads.ADSError as e:
+            raised_error = str(e)
+
+        # check no error was raised:
+        self.assertTrue(
+            not raised_error, f"Closing server connection raised: {raised_error}"
+        )
```

Comment:
If you could be bothered, I find something like:

```python
test_server.stop()
time.sleep(0.1)  # Give server a moment

try:
    test_int.__del__()
except ExceptionType as err:
    self.fail(f"Closing server connection raised: {raised_error}")
```
more neat (see [SO](https://stackoverflow.com/questions/4319825/python-unittest-opposite-of-assertraises)).

Reply:
Sure. I would be happy to consider your suggestions. Just to be clear: do you mean by the comment that you find the comments too cluttering, or do you refer to the `self.assertTrue(...)` check, or both?

Reply:
No, your in-code comments are fine. I mean the way of asserting an exception did _not_ get raised. I think the `except self.fail(...)` is cleaner than an `self.assert...` after the clause.

Reply:
Thank you for pointing it out and for the clarification. I changed it now with the recent commit [cbff3e8](https://github.com/stlehmann/pyads/pull/369/commits/cbff3e8975b80d2e478073478288c87851a52476).